lib/lhm/progress.rb

```diff
@@ -0,0 +1,202 @@
+require "lhm/printer"
+require "json"
+
+module Lhm
+  STATE_COPYING = "copying"
+  STATE_COPYING_DONE = "copying_done"
+  STATE_COPYING_FAILED = "copying_failed"
+  STATE_INITIAL = "initial"
+  STATE_SETUP_FAILED = "setup_failed"
+  STATE_SETUP_DONE = "setup_done"
+  STATE_SWITCHING_TABLES = "switching_tables"
+  STATE_SWITCHED_TABLES = "switched_tables"
+  STATE_SWITCHING_TABLES_FAILED = "switching_tables_failed"
+  STATE_TRIGGERS_DROPPED = "trigggers_dropeed"
+
+  class Progress
+    def initialize(origin, connection = nil, options: {})
+      @origin = origin
+      @connection = connection
+      @options = options
+      @throttler = options.include?(:throttler) ? options[:throttler].class.name : "None"
+      @table_switcher = options.include?(:atomic_switch) ? "AtomicSwitcher" : "LockedSwitcher"
+      @state = Lhm::STATE_INITIAL
+      @printer = options[:printer] || Printer::Percentage.new
+  
+      @rows_written = 0    
+      @speedometer_window = options[:speedometer_window] || 5 * 60
+      @speedometer = Speedometer.new(@speedometer_window)
+      @copy_speed = 0.0
+      @bytes_copied = 0
+      @avg_row_length = 0
+      @total_bytes = 0
+      @completion_percentage = 0
+      @start_time = Time.now
+
+      @retry_helper = SqlRetry.new(
+        @connection,
+        {
+          log_prefix: "Progress"
+        }.merge!(options.fetch(:retriable, {}))
+      )
+    end
+
+    attr_reader :state
+
+    def update_before_copy(start, limit)
+      @min_pk_key = start
+      @max_pk_key = limit
+    end
+
+    def update_during_copy(affected_rows)
+      @rows_written += affected_rows
+
+      @bytes_copied = @rows_written * @avg_row_length
+      @completion_percentage = ((@bytes_copied.to_f / @total_bytes) * 100.00).round(2)
```

Comment:
So if we do query information_schema every time, does it cause a lot of slowdowns due to additional queries? This might be a problem, and there are solutions to that problem. That said, I think we should probably start with the approach where we query information_schema directly, as opposed to tracking this by integrating/summing.

Reply:
The reason I am not querying information such as `bytes_written` or `avg_row_lenght` continously from the information_schema is not due to downtime but due to the fact that information_schema does not always return accurate information regarding tables stats, as it `caches` the information. 

In my initial version of this PR I had the logic to continously get the table stats (such as bytes_copied) while copying from the information_schema. While manually testing I found some glaring issues.

I had a table of apprx 100,000 rows and started Lhm with a chunk_size of 2000. During the entire course of the Lhm the bytes_copied on the shadow table was same due to the fact that information_schema caches values and does not directly update the stats of a table in use. This maybe a bit deceiving and error-prone to the end user. 

I read on mysql 5.7 docs that in order to update the table stats in information_schema we have to run a query like `ANALYZE_TABLE` (https://dev.mysql.com/doc/refman/5.7/en/analyze-table.html). So in order to get accurate information while copying we need to run two queries
```
ANALYZE TABLE 
SELECT * FROM INFORMATION_SCHEMA.TABLES .....
```

And reading from the mysql 5.7 docs it seems like `ANALYZE_TABLE` is a bit cumbersome query where it requires a flush lock. This is the wording from the docs 
> ANALYZE TABLE removes the table from the table definition cache, which requires a flush lock. If there are long running statements or transactions still using the table, subsequent statements and transactions must wait for those operations to finish before the flush lock is released. Because ANALYZE TABLE itself typically finishes quickly, it may not be apparent that delayed transactions or statements involving the same table are due to the remaining flush lock.

I don't have much context on whether always running `ANALYZE_TABLE` is safe, so just went with my intuition. Also would be happy to know if there's another way to fix this, or else we can just track the information by summing/integrating.